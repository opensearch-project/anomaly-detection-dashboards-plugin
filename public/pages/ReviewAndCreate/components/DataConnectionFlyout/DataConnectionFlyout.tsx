/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import {
  EuiButton,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFlyout,
  EuiFlyoutBody,
  EuiFlyoutFooter,
  EuiFlyoutHeader,
  EuiText,
  EuiLink,
  EuiSteps,
  EuiBasicTable,
  EuiTitle,
} from '@elastic/eui';
import React from 'react';

type DataConnectionFlyoutProps = {
  indices: any;
  onClose(): void;
  localClusterName: string;
};

const columns = [
  {
    field: 'cluster',
    name: 'Data connection',
    sortable: true,
    truncateText: true,
  },
  {
    field: 'index',
    name: 'Index',
    sortable: true,
    truncateText: true,
  },
];

const getIndexItemsToDisplay = (props: DataConnectionFlyoutProps) => {
  const indexItems = props.indices.map((index = '', int) => {
    const item = { id: int };
    const shouldSplit = index.includes(':');
    const splitIndex = index.split(':');
    let clusterName = shouldSplit ? splitIndex[0] : props.localClusterName;
    clusterName =
      clusterName === props.localClusterName
        ? `${clusterName} (Local)`
        : `${clusterName} (Remote)`;
    const indexName = shouldSplit ? splitIndex[1] : index;
    item.cluster = clusterName;
    item.index = indexName;
    return item;
  });
  return indexItems;
};

export const DataConnectionFlyout = (props: DataConnectionFlyoutProps) => {
  return (
    <EuiFlyout size="m" onClose={props.onClose}>
      <EuiFlyoutHeader hasBorder className="flyout">
        <EuiFlexGroup justifyContent="flexStart" alignItems="center">
          <EuiFlexItem className={'eui-textTruncate'}>
            <EuiTitle
              className={'eui-textTruncate'}
              size={'m'}
              data-test-subj={'dataSourcesFlyout_header'}
            >
              <h3>{`Index`}</h3>
            </EuiTitle>
          </EuiFlexItem>
          <EuiFlexItem grow={false}></EuiFlexItem>
        </EuiFlexGroup>
      </EuiFlyoutHeader>
      <EuiFlyoutBody
        // @ts-ignore
        style={{ overflowY: 'auto' }}
      >
        <EuiBasicTable
          items={getIndexItemsToDisplay(props)}
          itemId={(item) => item.id}
          columns={columns}
          //pagination={true}
          isSelectable={false}
          hasActions={false}
          noItemsMessage={'No data sources configured for this detector.'}
          data-test-subj={'dataConnectionsFlyout_table'}
        />
      </EuiFlyoutBody>
      <EuiFlyoutFooter>
        <EuiFlexGroup alignItems="center" justifyContent="flexEnd">
          <EuiFlexItem grow={true}></EuiFlexItem>
        </EuiFlexGroup>
      </EuiFlyoutFooter>
    </EuiFlyout>
  );
};
