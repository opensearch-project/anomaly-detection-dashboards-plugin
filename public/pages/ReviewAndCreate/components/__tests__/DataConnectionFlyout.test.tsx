/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import { render } from '@testing-library/react';
import { DataConnectionFlyout } from '../DataConnectionFlyout';

const renderDataConnectionFlyout = (
  indices: string[],
  localClusterName: string = 'local-cluster',
  onClose = jest.fn()
) => {
  return render(
    <DataConnectionFlyout
      indices={indices}
      localClusterName={localClusterName}
      onClose={onClose}
    />
  );
};

describe('<DataConnectionFlyout /> spec', () => {
  test('renders the flyout with indices and local cluster name', () => {
    const indices = [
      'test-index',
      'cluster-2:test-index-2',
      'cluster-3:http-index',
    ];
    const { container, getByText, getByTestId } =
      renderDataConnectionFlyout(indices);
    expect(container.firstChild).toMatchSnapshot();
    getByTestId('dataSourcesFlyout_header');
    getByText('local-cluster (Local)');
    getByText('test-index');
    getByText('cluster-2 (Remote)');
    getByText('test-index-2');
    getByText('cluster-3 (Remote)');
    getByText('http-index');
  });
  test('renders the flyout with only local indices', () => {
    const indices = ['local-index'];
    const { container, getByText, getByTestId, queryByText } =
      renderDataConnectionFlyout(indices);
    expect(container.firstChild).toMatchSnapshot();
    getByTestId('dataSourcesFlyout_header');
    getByText('local-cluster (Local)');
    getByText('local-index');
    expect(queryByText('Remote')).toBeNull();
  });

  test('handles no indices', () => {
    const indices: string[] = [];
    const { getByText, getByTestId } = renderDataConnectionFlyout(indices);

    getByTestId('dataSourcesFlyout_header');
    getByText('No data sources configured for this detector.');
  });

  test('calls onClose when flyout close button is clicked', () => {
    const onClose = jest.fn();
    const indices = ['test-index'];
    const { getByTestId } = renderDataConnectionFlyout(
      indices,
      'local-cluster',
      onClose
    );

    const closeButton = getByTestId('euiFlyoutCloseButton'); // Adjust the text if needed for close button
    closeButton.click();

    expect(onClose).toHaveBeenCalled();
  });
});
