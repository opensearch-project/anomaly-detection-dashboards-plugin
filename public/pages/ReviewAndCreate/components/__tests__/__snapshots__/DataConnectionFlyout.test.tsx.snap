// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DataConnectionFlyout /> spec renders the flyout with indices and local cluster name 1`] = `null`;

exports[`<DataConnectionFlyout /> spec renders the flyout with only local indices 1`] = `null`;
